app/models/application_record.rb

```diff
@@ -87,6 +87,22 @@ def self.aggregate_of(association)
     end
   end
 
+  def self.with_temporary_token(field_name, duration = 2.hours)
+    this = self
+    class_eval do
+      token_attribute = field_name
+      token_date_attribute = "#{field_name}_expiration_date"
+
+      define_method("generate_#{field_name}!") do
+        update!(token_attribute => this.generate_secure_token, token_date_attribute => duration.from_now)
```

Comment:
We usually use self.class idiom instead of saving the class in a variable

Reply:
But `self` within the closure is the destination object of the reflection, not the module. Any ideas?

Reply:
Sorry, I didn't follow you. `self` outside `class_eval` should evaluate to the destination object (User class in this case) and using `self.class` inside that method definition would evaluate to User class as well. And that's OK because the way you've defined `generate_secure_token` would make it available in that class.

Maybe I'm missing your point though :thinking: